### PR TITLE
Implementation of SAC

### PR DESCRIPTION
**Pull request by frederikschubert**
Since SAC is really similar to TD3, we are able to re-use most of its components. The differences are:

- The actions to update the q-functions and policy are sampled using the current policy (instead of taking the mode).
- There is no target policy.
- The log variance of the policy depends on the state.
- The policy is entropy regularized.

The current implementation does not support multi-step td-learning.

### CODE REVIEW

**Comment by frederikschubert** [comment]
Maybe we can use the `ClippedDoubleQLearning` and add a parameter whether the action should be sampled from the target policy or if the mode should be taken?

**Comment by KristianHolsheimer** [comment]
Wow you're fast!

> Maybe we can use the `ClippedDoubleQLearning` (instead of introducing a "Soft" version that is basically the same) and add a parameter whether the action should be sampled from the target policy or if the mode should be taken?

Yes, that would probably be cleaner.

File: coax/policy_objectives/_soft_pg.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# ------------------------------------------------------------------------------------------------ #
+
+import jax.numpy as jnp
+import haiku as hk
+import chex
+
+from ._deterministic_pg import DeterministicPG
+
+
+class SoftPG(DeterministicPG):
```
Comment (KristianHolsheimer) [inline coax/policy_objectives/_soft_pg.py:29]: Looks good!

File: doc/examples/pendulum/td3.rst [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Pendulum with DDPG
+Pendulum with TD3
```
Comment (KristianHolsheimer) [inline doc/examples/pendulum/td3.rst:1]: Thanks!

**Review by KristianHolsheimer — COMMENTED** [review]
I'm not sure which is better to add `SoftClipped...` or to add a flag to the existing `Clipped...`.

I personally prefer the not to have too many arguments exposed to the user. Perhaps it's worth the duplication. I don't feel strongly one way or another. Do what you think is best for new users.

**Comment by frederikschubert** [comment]
> I personally prefer the not to have too many arguments exposed to the user. Perhaps it's worth the duplication. I don't feel strongly one way or another. Do what you think is best for new users.

Yes, I also went with the duplication because it better separates both methods conceptually and there probably won't be many complicated refactorings that can't be applied easily to the duplicated code. I will leave a comment though as a pointer.

**Review by KristianHolsheimer — APPROVED** [review]

**Pull request merged**